### DISCUSSION TITLE
Help getting started with analog plugin

### DISCUSSION THREAD

**Discussion by fitch22** [grblHAL/core#96] [Q&A]
Hello,
I have a spindle that needs 0-10V for speed control.  I have tried a couple of the cheap pwm to 10V converters, and they are pretty miserable.  I had an idea that my board (BTT SKR 2) could output an analog voltage when it updates its pwm output, and I can just amplify the analog output to span the 0-10V range I need.  It would be linear and much more accurate than the output of a pwm converter.

So, I looked at the plugin examples and the templates but am still a bit lost.  Seems to me I need to:
1.  Initialize the proper output pin to be DAC output
2.  Intercept the execution of M3 so that when pwm duty cycle is updated that the DAC output gets updated, too
3.  Pay attention to some of the settings $30 to $33

Any tips on how to get started would be appreciated.

Thanks.

**Comment by terjeio** [comment]
1. Do that in you plugin_init function.
2. [Here](https://github.com/grblHAL/Plugins_spindle/blob/908a437271cb7fb8bb5b19bcfcfc9607f976d689/huanyang.c#L361-L368) is an example of how a plugin hooks into the HAL to provide alternative spindle control.
3. You can use the [core scaling functions](https://github.com/grblHAL/core/blob/637fa80eb328b042754716969674caff598425d9/spindle_control.c#L168-L229) for this?

**Reply by fitch22** [reply]
Terje,
I am stuck.

I added some code in my board_init() to enable an analog output pin and to replace hal.spindle.set_state with my_set_state().  my_set_state() takes care of setting the DAC value then calls the previous hal.spindle.set_state which takes care of all the existing pwm stuff.

Problem is, after my board_init() is called, a call to hal.settings_changed() is made (driver.c line 1313) and in line 1343 hal.spindle.set_state is overwritten.

Can I intercept settings_changed() somehow so that I can swap out my_set_state() later in the game?  Is there a better way to do this?

Thanks,
Bill

**Reply by terjeio** [reply]
> Can I intercept settings_changed() somehow so that I can swap out my_set_state() later in the game? 

You have to. See the [init function](https://github.com/grblHAL/Plugins_spindle/blob/908a437271cb7fb8bb5b19bcfcfc9607f976d689/huanyang.c#L381-L385) in the example I linked to above. The example allows switching between the VFD and PWM output on a settings change so is a bit more complex than you need it to be. It is important you call the saved settings changed function before reattaching your functions.

[Here](https://github.com/grblHAL/Plugin_plasma/blob/74210ad3216fd38ebeac4b89177898d548f93a22/thc.c#L394-L416) is another example - since this plugin has its own settings attaching to `hal.settings_changed` is delayed until they have been [loaded](https://github.com/grblHAL/Plugin_plasma/blob/74210ad3216fd38ebeac4b89177898d548f93a22/thc.c#L571-L572).

**Reply by fitch22** [reply]
Terje,
OK, got things working.  Thanks a million for your help.

In the process, I had to uncomment a line in stm32f4xx_hal_conf.h to define HAL_DAC_MODULE_ENABLED and I need to add the files for the dac in the drivers folder.  Should I put some conditional stuff to only enable the HAL_DAC_MODULE_ENABLED if it is a BTT_SKR_2.0 board?  Is there another place I can set this define where it won't conflict with other F4 boards?

Thanks,
Bill

**Reply by terjeio** [reply]
>  Is there another place I can set this define where it won't conflict with other F4 boards?

Just add 
`#include "stm32f4xx_hal_dac.h"`
to your plugin and add the generated code in `MX_DAC_Init()` in _main.c_ to your init function.

If doing it this way there will be no conflicts.

**Reply by fitch22** [reply]
Terje,
Thanks for your patience with me.  I am sorry to be such a pest...

If I put the include in my btt_skr_2.0.c file, I get the following error:
d:\st\stm32cubeide_1.6.1\stm32cubeide\plugins\com.st.stm32cube.ide.mcu.externaltools.gnu-tools-for-stm32.9-2020-q2-update.win32_2.0.0.202105311346\tools\arm-none-eabi\bin\ld.exe: ./Src/btt_skr_2.0.o: in function `board_init':
btt_skr_2.0.c:(.text.board_init+0x16): undefined reference to `HAL_DAC_Init'
d:\st\stm32cubeide_1.6.1\stm32cubeide\plugins\com.st.stm32cube.ide.mcu.externaltools.gnu-tools-for-stm32.9-2020-q2-update.win32_2.0.0.202105311346\tools\arm-none-eabi\bin\ld.exe: btt_skr_2.0.c:(.text.board_init+0x2a): undefined reference to `HAL_DAC_ConfigChannel'
collect2.exe: error: ld returned 1 exit status

HAL_DAC_Init prototype is in stm32f4xx_hal_dac.h, so I don't understand why it is not found.

If I uncomment the line in stm32f4xx_hal_conf.h to define HAL_DAC_MODULE_ENABLED then the linker finds everything and it builds.
Next I tried adding #define HAL_DAC_MODULE_ENABLED to my btt_skr_2.0.c file as well, but I end up with the same error.

I don't understand exactly what the failure is.  The symbol HAL_DAC_MODULE_ENABLED only shows up in 3 places in the entire project, in the 2 header files stm32f4xx_hal_dac.h and stm32f4xx_hal_dac_ex.h and the stm32f4xx_hal_conf.h.  Inside the Cube IDE hovering over the HAL_DAC_Init function call shows me the info about it and I can navigate to it.

Any ideas?
Bill

**Reply by terjeio** [reply]
> Any ideas?

Add `HAL_DAC_MODULE_ENABLED`  to the build configuration?

![image](https://user-images.githubusercontent.com/20260062/144762382-0b30074e-b42e-49a8-aa36-575925979c33.png)

**Reply by fitch22** [reply]
That works.  Thanks!
